packages/ts/generator-plugin-signals/src/SignalProcessor.ts

```diff
@@ -44,30 +46,48 @@ export default class SignalProcessor {
     const [file] = ts.transform<SourceFile>(this.#sourceFile, [
       transform((tsNode) => {
         if (ts.isFunctionDeclaration(tsNode) && tsNode.name && this.#methods.has(tsNode.name.text)) {
-          const filteredParams = tsNode.parameters.filter(
-            (p) => !p.type || !ts.isTypeReferenceNode(p.type) || p.type.typeName !== initTypeId,
-          );
-          if (filteredParams.length > 0) {
-            functionParams.set(tsNode.name.text, filteredParams);
-          }
-          const paramNames = filteredParams.map((p) => (p.name as ts.Identifier).text).join(', ');
           const signalId = this.#replaceSignalImport(tsNode);
           let genericReturnType;
           if (genericSignals.includes(signalId.text)) {
             genericReturnType = (tsNode.type as ts.TypeReferenceNode).typeArguments![0];
           }
           const returnType = genericReturnType ?? signalId;
-          const INITIAL_VALUE = signalId.text.startsWith('NumberSignal') ? '0' : 'undefined';
+          let initialValue: ts.Expression = signalId.text.startsWith('NumberSignal')
+            ? ts.factory.createNumericLiteral('0')
+            : ts.factory.createIdentifier('undefined');
+          const filteredParams = tsNode.parameters.filter(
+            (p) => !p.type || !ts.isTypeReferenceNode(p.type) || p.type.typeName !== initTypeId,
+          );
+          // `filteredParams` can be altered after, need to store the param names now
+          const paramNames = filteredParams.map((p) => (p.name as ts.Identifier).text).join(', ');
+          // Searching for a non-nullable signal like ValueSignal<SomeType>
+          if (
+            ts.isUnionTypeNode(returnType) &&
+            returnType.types.length &&
+            ts.isTypeReferenceNode(returnType.types[0]) &&
+            returnType.types[0].typeArguments?.length === 1 &&
+            ts.isUnionTypeNode(returnType.types[0].typeArguments[0]) &&
+            returnType.types[0].typeArguments[0].types.length === 1
+          ) {
+            const requiredDefaultType = returnType.types[0].typeArguments[0].types[0];
```

Comment:
I'd suggest to wrap this logic in a function called something like `getRequiredDefaultValueType` for the sake of better readability.